### PR TITLE
docs/design: Initial diagnostics design

### PR DESCRIPTION
**Pull request by bflad**
Reference: https://github.com/hashicorp/terraform-plugin-framework/issues/24

### CODE REVIEW

File: docs/design/diagnostics.md [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+// This could be folded into Diagnostic
+type DiagnosticWithPath interface {
+  Diagnostic
+
+  DiagnosticPath() *tftypes.AttributePath // See also: https://github.com/hashicorp/terraform-plugin-framework/issues/81
+}
+```
+
+This would allow a single diagnostic to be usable in place of where an `error` would be returned and the type would pick up associated `errors` package capabilities. All implementations would be required for including an `Error()` method and optionally other `Unwrap()` etc. methods to satisfy other `errors` capabilities, which could introduce verbosity concerns. It also might be unclear what the semantics of a warning diagnostic are in terms of the `error` interface -- this implementation would treat all warning diagnostics as errors, which seems less than ideal.
```
Comment (apparentlymart) [inline docs/design/diagnostics.md:289]: There's some prior art related to this in both HCL and in Terraform, which I wanted to note in case it's useful for thinking through the tradeoffs here. I'm not intending this as any recommendation to do anything in particular, because there are likely some different requirements and preferences at the SDK layer than at these two layers, which you all are in a far better place to mull than I am.

* HCL's `hcl.Diagnostics` implements `error` in a way similar to what's described here, but it's at the level of the group of diagnostics rather than at the level of a single diagnostic so that a group of diagnostics can return together as a single `error` value.

    While this worked okay for the internals of HCL where everything was generally in agreement about how diagnostics work, it ended up causing friction once we started using diagnostics in Terraform where there's more of a blend of native diagnostics handling and traditional `error` handling. In particular, we got caught out a few times by the fact that a `hcl.Diagnostics` which contains only `hcl.DiagWarning` diagnostics appears as a non-`nil` `error`, even though it's not actually describing an error condition, which I think is the same concern raised in the paragraph above.

* In response to those problems, we created [Terraform's own `tfdiags` package](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags), which has [`tfdiags.Diagnostics`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags#Diagnostics) as a more general analog to `hcl.Diagnostics`. Part of that design was to intentionally make `tfdiags.Diagnostics` _not_ implement `error`, and instead it has a method [`Err`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags#Diagnostics.Err) which returns an `error` which is `nil` unless `diags.HasErrors()`, thus preserving the usual meaning of a non-`nil` error at the expense of then losing track of warnings that aren't accompanied by an error (because there's nowhere to put them in an `error` value).

    For some particularly gnarly cases we also have [`ErrWithWarnings`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags#Diagnostics.ErrWithWarnings) which essentially recovers the HCL approach of potentially returning a weird `error` that might actually only be reporting warnings. In the few cases where we use that the caller needs to be careful to check whether the `error` value has the type [`NonFatalError`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags#NonFatalError) and treat it as warnings only in that case.

* Tangentially, we also made [the `Append` method of `Diagnostics`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags#Diagnostics) able to accept naked `error` values and turn them into (often lower-quality) diagnostic errors. It also recognizes `error` values previously returned from `Diagnostics.Err` or `Diagnostics.NonFatalErr` and recovers the original diagnostics from them, which allows losslessly(-ish) sending diagnostics through a function that returns `error` because e.g. it needs to interact with other typical Go error-handling patterns.

None of these designs are ideal by any means, but the `tfdiags.Diagnostics` design has been working reasonably well and I personally think it was an improvement over the `hcl.Diagnostics` design in terms of it being easier to use it correctly.

Comment (bflad) [reply docs/design/diagnostics.md:289]: Thank you for this additional context, @apparentlymart! I'm wondering if it might be good to capture some of this related context into the design documentation in the prior art section for posterity. 🤔 

Given the various complexities mentioned, it makes it feel more pertinent to treat diagnostics as explicitly different than regular `error`s to reduce cognitive load. We don't have a use case for going from a diagnostic to an `error` currently, which probably simplifies our implementation. We do have the use case for promoting an `error` to a diagnostic (especially outside the framework since that is how vendor SDKs are generally written), but maybe requiring framework and provider developers work through converting them explicitly is a good nudge for writing better messaging rather than just writing the helper function to wholesale move the error string into an error diagnostic summary.

File: docs/design/diagnostics.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+Diagnostics in this framework should be:
+
+- Able to natively support warning and error diagnostics.
+- Implementable by provider developers.
+- Abstracted from terraform-plugin-go and convertable into those types to separate implementation concerns.
+- Ergonomic to create, read, and delete in Go code (e.g. have helper methods for common use cases).
+
+Additional consideration should be given to:
+
+- Extensibility and additional data storage (e.g. distinguishing between types of diagnostics or what generated an error).
```
Comment (paddycarver) [inline docs/design/diagnostics.md:138]: So I think this covers a way to achieve what we want, but I think what we want is to be able to handle diagnostics programmatically, whereas right now they're mostly tailored for a human, practitioner audience.

Also, could you expand on "what generated an error" means here?

File: docs/design/diagnostics.md [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+type Diagnostics []Diagnostic
+
+func (d Diagnostics) Append(in ...Diagnostic)
+func (d Diagnostics) HasError() bool
+func (d Diagnostics) toTfprotov6Diagnostics() []*tfprotov6.Diagnostic
+```
+
+This solves many requirements, allowing the framework to own the abstraction and provide ergonomic methods. However, this implementation is tightly coupled to the underlying implementation and does not introduce protections against backwards compatibility issues. Since the data fields are strongly associated with the type, implementors inside and outside the framework are bound to the initial implementation details. If those details are not exported and instead hidden behind a constructor function, e.g.
```
Comment (paddycarver) [inline docs/design/diagnostics.md:164]: Can you expand on _why_ it would be difficult to change? Elsewhere (e.g., request/response pattern) we've argued that using structs _helps_ backwards compatibility. What sets this apart?

File: docs/design/diagnostics.md [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    detail    string
+    attribute *tftypes.AttributePath // See also: https://github.com/hashicorp/terraform-plugin-framework/issues/81
+}
+
+func NewDiagnostic(severity DiagnosticSeverity, summary string, detail string, attribute *tftypes.AttributePath) Diagnostic
+func NewErrorDiagnostic(summary string, detail string, attribute *tftypes.AttributePath) Diagnostic
+func NewWarningDiagnostic(summary string, detail string, attribute *tftypes.AttributePath) Diagnostic
+```
+
+This type of implementation will still be difficult to extend or change.
```
Comment (paddycarver) [inline docs/design/diagnostics.md:179]: Why?

File: docs/design/diagnostics.md [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+As another extensibility example, as `Diagnostics` pass through various functions across subsystems, determining the "source" or "cause" of a diagnostic must be done by `Summary` and/or `Detail` value similar to checking Go `error` types by value using `(error).Is()`. Extending the `Diagnostic` type is possible to add other identifying fields or creating new types, e.g.
```
Comment (paddycarver) [inline docs/design/diagnostics.md:181]: > determining the "source" or "cause" of a diagnostic must be done by `Summary` and/or `Detail` value similar to checking Go `error` types by value using `(error).Is()`

Would this abstraction allow for that? I'm not sure how a `Diagnostic.Is` would work in this case.

File: docs/design/diagnostics.md [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+```go
+type ValidationDiagnostic struct {
+  Diagnostic
+}
+```
```
Comment (paddycarver) [inline docs/design/diagnostics.md:187]: I don't know that this would work. I think we'd get compilation errors on this:

```go
func doTheThing() Diagnostic {
  return ValidationDiagnostic{}
}
```

I don't think that would work 🤔 I think that kind of embedding only works for interfaces; the struct embedding I think just inherits properties and methods, but the two types can't be used interchangeably. (I'd love to be corrected if I'm wrong, or if I'm misunderstanding what you're suggesting here.)

File: docs/design/diagnostics.md [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+  Detail()   string
+
+  Equal(other Diagnostic) bool
+}
+
+type DiagnosticWithLogger interface {
+  Diagnostic
+
+  Log(ctx context.Context)
+}
```
Comment (paddycarver) [inline docs/design/diagnostics.md:210]: Unclear on what this interface is or what it would do.

File: docs/design/diagnostics.md [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+  if !ok {
+    return false
+  }
+
+  return other.Summary() == d.Summary() && other.Detail() == d.Detail()
+}
+```
+
+It then becomes possible to distinguish between diagnostic types (in this case the "source" of a diagnostic) while still supporting additional functionality defined by the interface. These extended diagnostics can also implement other framework-defined capabilities, such as implementing the example `Log()` method in this situation.
```
Comment (paddycarver) [inline docs/design/diagnostics.md:264]: I think the use of "source" is throwing me, because I keep expecting it to be able to work like a stacktrace, helping me find the line a diagnostic was returned from. But I think the comment about distinguishing between the type of diagnostics is really illuminating.

**Review by paddycarver — CHANGES_REQUESTED** [review]
I like the proposal, I agree with the recommendation at the end, and I think you did a really great job with this document. Amazing work.

I think the part I'm not seeing explicitly called out here that we talked about informally is your idea of validation helpers like `Not`, and how the different approaches intersect with that kind of work.

I think as we make diagnostics more prevalent throughout the codebase and require provider devs to interface with them more often, we're going to run into more and more situations where we're going to want to discriminate between diagnostics using more than just their severity, that different _types_ or _kinds_ of diagnostics will merit different handling. And I think your proposal will do a good job of addressing that.

I left some comments asking for clarification on things, but I think the proposal is good, I'm mostly just asking so we can surface and record context.

File: docs/design/diagnostics.md [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    message Request {
+        DynamicValue config = 1;
+    }
+    message Response {
+        repeated Diagnostic diagnostics = 2;
+    }
+}
+```
+
+All RPCs except `StopProvider` include diagnostics support.
```
[the remark this answers is not in the document]
Comment (bflad) [reply docs/design/diagnostics.md]: Added a quick mention: https://github.com/hashicorp/terraform-plugin-framework/pull/108/commits/7b45926d3ce28508e48cb4d54ee823535ebfda98 -- if there's something else expected here, please let me know.

File: docs/design/diagnostics.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+
+func FromErr(err error) Diagnostics
+```
+
+## Goals
+
+Diagnostics in this framework should be:
+
+- Able to natively support warning and error diagnostics.
+- Implementable by provider developers.
```
[the remark this answers is not in the document]
Comment (bflad) [reply docs/design/diagnostics.md:146]: We could make diagnostics unexported types in the framework, which makes them unusable. I'll reword this to be clearer that this specifically means providing exported types.

File: docs/design/diagnostics.md (hunk 2 above)
Comment (bflad) [reply docs/design/diagnostics.md]: I like focusing on the word "programmatic" here -- hopefully clarified these with https://github.com/hashicorp/terraform-plugin-framework/pull/108/commits/a29e9cda0d7fd3c5e145103e5986721c7b3eab3d

File: docs/design/diagnostics.md (hunk 5 above)
Comment (bflad) [reply docs/design/diagnostics.md:196]: I'll just remove the second half of that sentence.

File: docs/design/diagnostics.md (hunk 6 above)
Comment (bflad) [reply docs/design/diagnostics.md:202]: Fixed, thanks for calling that out.

File: docs/design/diagnostics.md (hunk 7 above)
Comment (bflad) [reply docs/design/diagnostics.md:227]: It's meant as an example extension, where the framework could support writing a log entry matching a diagnostic. Will add code comments, or happy to show a different example.

File: docs/design/diagnostics.md (hunk 3 above)
Comment (bflad) [reply docs/design/diagnostics.md:179]: Removed 👍

File: docs/design/diagnostics.md (hunk 4 above)
Comment (bflad) [reply docs/design/diagnostics.md:194]: Reworded 👍

**Review by paddycarver — APPROVED** [review]
This has everything I can think of in it. Great work.

File: docs/design/diagnostics.md [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+There are many pieces of framework functionality that warrant enhanced practitioner feedback capabilities:
+
+- Provider (`Configure()`), managed resource (CRUD functions), and data source (`Read()`) logic
+- Framework defined provider development issues such as:
+  - Invalid schema implementations
+  - Invalid getting or setting of configuration, plan, or state values
+- [Plan Modification](./plan-modification.md)
+- [Validation](./validation.md)
+
+This design documentation will walkthrough and recommend options for abstracted diagnostics handling and extensibility in the framework.
```
Comment (kmoe) [inline docs/design/diagnostics.md:14]:
```suggestion
This design documentation will walk through and recommend options for abstracted diagnostics handling and extensibility in the framework.
```

File: docs/design/diagnostics.md [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+    message Request {
+        DynamicValue config = 1;
+    }
+    message Response {
+        repeated Diagnostic diagnostics = 2;
+    }
+}
+```
+
+All RPCs except `StopProvider`, which never surfaces practitioner errors, include diagnostics support.
```
Comment (kmoe) [inline docs/design/diagnostics.md:83]:
```suggestion
All RPCs except `StopProvider`, which never surfaces errors to the practitioner, include diagnostics support.
```

Or, "surfaces provider errors", perhaps?

**Review by kmoe — APPROVED** [review]
LGTM, with optional readability fixes
